src/transformers/modeling_utils.py

```diff
@@ -3116,21 +3121,55 @@ def from_pretrained(
             config_path = config if config is not None else pretrained_model_name_or_path
-            config, model_kwargs = cls.config_class.from_pretrained(
-                config_path,
-                cache_dir=cache_dir,
-                return_unused_kwargs=True,
-                force_download=force_download,
-                resume_download=resume_download,
-                proxies=proxies,
-                local_files_only=local_files_only,
-                token=token,
-                revision=revision,
-                subfolder=subfolder,
-                _from_auto=from_auto_class,
-                _from_pipeline=from_pipeline,
-                **kwargs,
-            )
+            if from_gguf is None:
+                config, model_kwargs = cls.config_class.from_pretrained(
+                    config_path,
+                    cache_dir=cache_dir,
+                    return_unused_kwargs=True,
+                    force_download=force_download,
+                    resume_download=resume_download,
+                    proxies=proxies,
+                    local_files_only=local_files_only,
+                    token=token,
+                    revision=revision,
+                    subfolder=subfolder,
+                    _from_auto=from_auto_class,
+                    _from_pipeline=from_pipeline,
+                    **kwargs,
+                )
+            else:
+                from .modeling_gguf_pytorch_utils import load_and_convert_gguf_config
+
+                if not is_gguf_available():
+                    raise ValueError(
+                        "You need to have `gguf` installed in order to convert GGUF weights. `pip install gguf`"
+                    )
+
+                # Case 1: the GGUF file is present locally
+                if os.path.isfile(from_gguf):
+                    gguf_path = from_gguf
+                # Case 2: The GGUF path is a location on the Hub
+                # Load from URL or cache if already cached
+                else:
+                    cached_file_kwargs = {
+                        "cache_dir": cache_dir,
+                        "force_download": force_download,
+                        "proxies": proxies,
+                        "resume_download": resume_download,
+                        "local_files_only": local_files_only,
+                        "token": token,
+                        "user_agent": user_agent,
+                        "revision": revision,
+                        "subfolder": subfolder,
+                        "_raise_exceptions_for_gated_repo": False,
+                        "_raise_exceptions_for_missing_entries": False,
+                        "_commit_hash": commit_hash,
+                    }
+
+                gguf_path = cached_file(pretrained_model_name_or_path, from_gguf, **cached_file_kwargs)
+
+                config = load_and_convert_gguf_config(gguf_path)
+                model_kwargs = kwargs
```

Comment:
I would likely send that all to another method to take care of that to not add too much code to the already bloated `from_pretrained`

Reply:
Makes sense! Done !